## Description:
Calculation of $F_{ST}$ for pairwise populations, following the instruction of the following paper    
> 1984-Estimating F‚Äêstatistics for the analysis of population structure

## Dependencies:
- python2.7
- argparse
- sys
- socket
- os
- time
- itertools
- numpy
- pandas
- subprocess
- concurrent
- math
- operator
- plink1.9


## Input:

1. bplink file `required`
2. sample info file `required`
3. population pairs `optional`
4. indicator for whether output $F_{ST}$ values of all the variants `optional`
5. target regions to be analyzed `optional`
6. number of threads `optional`

Note: "plink1.9" would be used to calculate allele frequenies    
Make sure that plink1.9 is installed and could be found in your PATH environment.      

## Usage:    
### $F_{ST}$ (site- / region-based, and global)
```shell
python2.7 Fst.py -h
```

```shell
python2.7 Fst.py \
    --bfile prefix_of_bplink_file \
    --info sample.info \
    --pair population.pair \
    --point N \ 
    --region one \
    --thread 5 \
    --out ./out 
```

**Details about the arguments**

**--bfile** (required): /bplink/file/prefix. Note: duplicated snpID or indel are not allowed; Same FID and IID in the fam file, with/without gender info.

**--info** (required): individual info file, only samples in the info file will be used. \<IID\> \<population ID\>, 2 columns, no header, space or tab-delimited.

**--pair** (optional): pairs of populations to be analyzed. 2 columns, <pop1> <pop2>, no header, space or tab-delimited. $F_{ST}$ calculated only for population pairs in the pair file, OR, pairwisely, according to the populations listed in the info file

**--point** (optional): whether to output the $F_{ST}$ value for all the variants, choice: Y / N. default: N, output only global $F_{ST}$ values.

**--region** (optional): region file. To calculate FST values within the given regions, rather than the whole data set. 4 columns: <region ID> <chrom ID> <start physical pos> <end physical pos>

**--thread** (optional): number of processes. default: 20

**--out** (optional): prefix of output files, default: ./out     

---

### $F_{ST}$ bootstrapping (by sampling sites)
```shell
python2.7 Fst_bootstrap.py -h
```

```shell
python2.7 Fst_bootstrap.py \
    --bfile prefix_of_bplink_file \
    --info sample.info \
    --freqs /path/to/frq.gz/files \
    --sites /path/to/site/list \
    --size 100000 \
    --sampling block \ 
    --boot 100 \
    --pair population.pair \
    --thread 5 \
    --out ./out 
```

**Details about the arguments**

**--bfile** (required): The same as required by Fst.py.

**--info** (required): The same as required by Fst.py.

**--freqs** (required): Directory for frq.gz files, generated while running Fst.py.

**--pair** (optional): The same as required by Fst.py.

**--sites** (optional): Directory for prepared resample.site.${n}.list.gz files, ${n} ranges from 1 to the number of bootstrapping, (default: will be generated by this script if not provided).

**--size** (optional): Block size (bp) for sites sampling if sites were not provided, default: 100000 .

**--sampling** (optional): Sampling strategy. choices: 'block','snp','NO'. NO: with site files provided; block: sample moving blocks; snp: randomly sample one SNP for each block. default: block

**--boot** (optional): Numbers of bootstrapping, default: 100

**--thread** (optional): number of processes. default: 20

**--out** (optional): prefix of output files, default: ./out     

---
By: Yuwen Pan, 2020  
Contact: panyuwen.x@gmail.com

